File: lintcode/python/0612_k_closest_points.py
#!/usr/bin/python -t


"""
Definition for a point.
class Point:
    def __init__(self, a=0, b=0):
        self.x = a
        self.y = b
"""

import heapq

class Solution:
    """
    @param points: a list of points
    @param origin: a point
    @param k: An integer
    @return: the k closest points
    """
    def kClosest(self, points, origin, k):
        # write your code here
        hq = []
        for p in points:
            dist = self.get_dist(p, origin)
            heapq.heappush(hq, (-dist, -p.x, -p.y))
            if len(hq) > k:
                heapq.heappop(hq)

        ret = []
        while len(hq) > 0:
            dist, x, y = heapq.heappop(hq)
            ret.append(Point(-x, -y))

        #print(ret)
        ret.reverse()

        return ret

    def get_dist(self, p, o):
        return (p.x - o.x)**2 + (p.y-o.y)**2

# heap
# 基于 PriorityQueue 的方法
# PriorityQueue 里从远到近排序。当 PQ 里超过 k 个的时候，就 pop 掉一个。
# 时间复杂度 O(nlogk)O(nlogk)
# 
# 如果使用 Quick Select 离线算法：
# 0. 计算所有的点到原点的 distance -- O(n)
# 
# Quick Select 去找到 kth smallest distance -- O(n)
# 遍历所有 distance 找到 top k smallest distance -- O(n)
# 找到的 top k smallest points 按 distance 排序并返回 -- O(klogk)
# 总共 Quick Select 离线算法耗费时间 O(n + klogk)O(n+klogk)

"""
Definition for a point.
class Point:
    def __init__(self, a=0, b=0):
        self.x = a
        self.y = b
"""

import heapq

class Solution:
    """
    @param points: a list of points
    @param origin: a point
    @param k: An integer
    @return: the k closest points
    """
    def kClosest(self, points, origin, k):
        # write your code here
        heap = []
        ret = []
        
        for point in points:
            dist = self.getDist(point, origin)
            heapq.heappush(heap, (-dist, -point.x, -point.y))
            
            if len(heap) > k:
                heapq.heappop(heap)
                
        while len(heap) > 0:
            dist, x, y = heapq.heappop(heap)
            ret.append(Point(-x, -y))
            
        ret.reverse()
        
        return ret
            

File: lintcode/golang/0612_k_closest_points.go

// heap and sort

/**
 * Definition for a point.
 * type Point struct {
 *     X, Y int
 * }
 */

import (
    "container/heap"
    "sort"
)

type Node struct {
	val int
	X   int
	Y   int
}

type IntHeap []*Node

func (h IntHeap) Len() int           { return len(h) }
func (h IntHeap) Less(i, j int) bool { return h[i].val < h[j].val || (h[i].val == h[j].val && h[i].X > h[j].X) || (h[i].val == h[j].val && h[i].X == h[j].X && h[i].Y > h[j].Y) }
func (h IntHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }

func (h *IntHeap) Push(x interface{}) {
	// Push and Pop use pointer receivers because they modify the slice's length,
	// not just its contents.
	*h = append(*h, x.(*Node))
}

func (h *IntHeap) Pop() interface{} {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[0 : n-1]
	return x
}
/**
 * @param points: a list of points
 * @param origin: a point
 * @param k: An integer
 * @return: the k closest points
 */
func kClosest (points []*Point, origin *Point, k int) []*Point {
    // write your code here
	maxHeap := &IntHeap{}

	for _, p := range points {
		dist := getDist(p, origin)
		heap.Push(maxHeap, &Node{val: -dist, X: p.X, Y: p.Y})
		if maxHeap.Len() > k {
			heap.Pop(maxHeap)
		}
	}

	ret := []*Point{}
    
	for maxHeap.Len() > 0 {
		node := heap.Pop(maxHeap).(*Node)
		ret = append([]*Point{&Point{X: node.X, Y: node.Y}}, ret...)
	}

    sort.Slice(ret, func(i, j int) bool {
		d1 := (ret[i].X-origin.X)*(ret[i].X-origin.X) + (ret[i].Y-origin.Y)*(ret[i].Y-origin.Y)
		d2 := (ret[j].X-origin.X)*(ret[j].X-origin.X) + (ret[j].Y-origin.Y)*(ret[j].Y-origin.Y)
		return d1 < d2 || (d1 == d2 && ret[i].X < ret[j].X) ||(d1 == d2 && ret[i].X == ret[j].X && ret[i].Y < ret[j].Y)
	})

	return ret
}

func getDist(p, o *Point) int {
    return (p.X-o.X)*(p.X-o.X) + (p.Y-o.Y)*(p.Y-o.Y)
}
